src/main/java/org/spongepowered/mod/text/selector/SpongeSelectorBuilder.java

```diff
@@ -0,0 +1,129 @@
+import org.spongepowered.api.text.selector.SelectorType;
+import org.spongepowered.api.text.selector.SelectorTypes;
+import org.spongepowered.api.util.annotation.NonnullByDefault;
+
+import com.flowpowered.math.vector.Vector3i;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Maps;
+
+@NonnullByDefault
+public class SpongeSelectorBuilder implements SelectorBuilder {
+
+    private SelectorType type;
+    private Map<String, String> arguments = Maps.newHashMap();
+
+    public SpongeSelectorBuilder() {
+    }
+
+    public SpongeSelectorBuilder(Selector copyFrom) {
+        this.type = copyFrom.getType();
+        this.arguments = copyFrom.getArguments();
+    }
+
+    @Override
+    public SelectorBuilder type(SelectorType type) {
+        this.type = type;
+        return this;
+    }
+
+    @Override
+    public SelectorBuilder addArgument(String key, String value) {
+        this.arguments.put(key, value);
+        return this;
+    }
+
+    @Override
+    public SelectorBuilder addArgument(String key, int value) {
+        addArgument(key, Integer.toString(value));
+        return this;
+    }
+
+    @Override
+    public SelectorBuilder removeArgument(String key) {
+        this.arguments.remove(key);
+        return this;
+    }
+
+    @Override
+    public SelectorBuilder center(Vector3i center) {
+        addArgument("x", center.getX());
+        addArgument("y", center.getY());
+        addArgument("z", center.getZ());
+        return this;
+    }
+
+    @Override
+    public SelectorBuilder name(String name) {
+        name(name, false);
+        return this;
+    }
+
+    @Override
+    public SelectorBuilder name(String name, boolean invert) {
+        addArgument("name", (invert ? "!" : "") + name);
+        return this;
+    }
+
+    @Override
+    public SelectorBuilder team(String team) {
+        team(team, false);
+        return this;
+    }
+
+    @Override
+    public SelectorBuilder team(String team, boolean invert) {
+        addArgument("team", (invert ? "!" : "") + team);
+        return this;
+    }
+
+    @Override
+    public SelectorBuilder type(EntityType type) {
+        type(type, false);
+        return this;
+    }
+
+    @Override
+    public SelectorBuilder type(EntityType type, boolean invert) {
+        addArgument("type", (invert ? "!" : "") + type.toString());
+        return this;
+    }
+
+    @Override
+    public Selector build() {
+        Preconditions.checkState(this.type != null);
+        boolean requiresLocation = this.arguments.containsKey("x") || this.arguments.containsKey("y") || this.arguments.containsKey("z") || this.arguments.containsKey("dx") || this.arguments.containsKey("dy") || this.arguments.containsKey("dz") || this.arguments.containsKey("rm") || this.arguments.containsKey("r") || this.type.equals(SelectorTypes.NEAREST_PLAYER) || (this.arguments.containsKey("c") && !this.type.equals(SelectorTypes.RANDOM_PLAYER));
```

Comment:
Should be:
```boolean requiresLocation = this.arguments.containsKey("x") || this.arguments.containsKey("y") || this.arguments.containsKey("z") || this.arguments.containsKey("dx") || this.arguments.containsKey("dy") || this.arguments.containsKey("dz") || this.arguments.containsKey("rm") || this.arguments.containsKey("r") || this.type.equals(SelectorTypes.NEAREST_PLAYER);```

`c` doesn't matter for location. I'm thinking of a better way to do this checking that's more DRY.